src/System.Text.JsonLab/System/Text/Json/Utf8JsonReader_MultiSegment.cs

```diff
@@ -145,17 +180,33 @@ private bool ReadMultiSegment(ref BufferReader<byte> reader)
 
             if (TokenType == JsonTokenType.StartObject)
             {
-                reader.Advance(1);
-                CurrentIndex++;
                 if (first == JsonConstants.CloseBrace)
                 {
+                    reader.Advance(1);
+                    CurrentIndex++;
+                    _position++;
                     EndObject();
                 }
                 else
                 {
-                    if (first != JsonConstants.Quote) JsonThrowHelper.ThrowJsonReaderException(ref this);
+                    if (first != JsonConstants.Quote)
+                        ThrowJsonReaderException(ref this);
                     TokenStartIndex++;
-                    ConsumePropertyName(ref reader);
+                    int prevCurrentIndex = CurrentIndex;
+                    int prevPosition = _position;
+                    BufferReader<byte> copy = reader;
+                    reader.Advance(1);
+                    CurrentIndex++;
+                    _position++;
```

Comment:
Can we somehow simplify all these indexes? It seems like we basically have 3 different ints tracking pretty much the same position (with different origins). Can we save the origins, and then only increment one integer?

Reply:
I have been thinking about this issue but haven't been able to come up with a solution that works for all cases. **I will see if I can figure something out.** The issue is the origin is only useful for subsequent calls in the outerloop. Within the loop, I need the concept of where a token starts (specifically strings). Generally, it isn't too expensive to track these. _position and CurrentIndex are too different to reconcile since new lines can essentially appear anywhere (especially when dealing with partially processed tokens).